Lock TT by holding finger on home screen

mmilata opened the pull request:
Fixes #1404. The usual hold-to-confirm animation is shown after first 500ms and finishes in 5s in total. The PR also extends debuglink to support touch hold events.

Please test on actual device as the overall impression can be very different from emulator:)
([regular debug build](https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/972529766/artifacts/file/trezor-fw-regular-debug-2.3.6-6ce0745b.bin))

![Peek 2021-01-19 16-47](https://user-images.githubusercontent.com/85857/105058011-15019c80-5a76-11eb-8d9d-0d8648d1bdc0.gif)

matejcik commented:
:100: for `hold_ms`, very nice!

matejcik commented:
playing around with this a little.

5 seconds total time seems way too high, for me personally 2.5s is about right. The 500 ms delay at start feels exactly right.

I've also experienced some artifacts:
- with the 5-second total time, the decay when you lift your finger is WAY too slow. But maybe we don't need to concern ourselves with that, if others agree to shorten the total time.
- the response between lifting finger and pressing it again is also slow. I'm not sure where the lag comes from though.
- after I lock the device this way, there's a screen refresh after the auto-lock delay elapses. I believe the same thing happens with the `LockDevice` message, and I'm honestly not sure what to do about it or whether it is even worth doing anything about; we'd need to either disengage the idle timer, or make sure repeated locking does not do screen refresh.
  OTOH I like that `lock_device` is unconditional in this regard, so maybe we should have a `lock_if_unlocked` function for this?

tsusanka commented:
Discussed with @matejzak: let's keep the design as is for now and we will remake the UI along with the other redesign. We also agree with @matejcik's time suggestions: 0.5s delay and 2.5s to lock.

mmilata commented:
Updated the time to 2.5s, agreed that it feels better now.

> with the 5-second total time, the decay when you lift your finger is WAY too slow. But maybe we don't need to concern ourselves with that, if others agree to shorten the total time.

The reverse speed is currently hardcoded to be 2x the forward one, so the reverse animation now takes 1.25s at most. Would be easy to make this parametric and use a factor of 3 or 4.

> after I lock the device this way, there's a screen refresh after the auto-lock delay elapses. I believe the same thing happens with the LockDevice message, and I'm honestly not sure what to do about it or whether it is even worth doing anything about; we'd need to either disengage the idle timer, or make sure repeated locking does not do screen refresh.
OTOH I like that lock_device is unconditional in this regard, so maybe we should have a lock_if_unlocked function for this?

Good catch! I've implemented the `lock_if_unlocked` suggestion in 56f8880, PTAL.

> the response between lifting finger and pressing it again is also slow. I'm not sure where the lag comes from though.

Not quite sure what you mean, can you please describe the sequence of actions needed to reproduce it?

mmilata commented on core/src/apps/base.py:
```diff
@@ -265,4 +270,4 @@ def boot() -> None:
 
     wire.experimental_enabled = storage.device.get_experimental_features()
 
-    workflow.idle_timer.set(storage.device.get_autolock_delay_ms(), lock_device)
+    workflow.idle_timer.set(storage.device.get_autolock_delay_ms(), lock_device_if_unlocked)
```
Perhaps we should call `reload_settings_from_storage` here?

mmilata replied:
Tiny bit of code duplication (which I missed the first time), we do the same two things there (plus setting display orientation which is idempotent). Dunno, probably not worth it for 2 lines.

matejcik replied:
oh I see now. yes, that would make sense, but we would have to move `reload_settings_from_storage` to `common` probably?

mmilata replied:
Moved it to `apps/base.py` in aa07da9dc7544056a73f89b586f9f797e5dabc79. Not sure where in `common` would it go, and some other apps import from `apps.base` already.

matejcik commented:
> > the response between lifting finger and pressing it again is also slow. I'm not sure where the lag comes from though.
> 
> Not quite sure what you mean, can you please describe the sequence of actions needed to reproduce it?

1. hold finger, wait until progress bar is 50% filled
2. lift finger briefly
3. hold again

expected behavior: progress resumes from the exact spot to which it decayed

actual behavior: progress continues to decay for a small while, and then starts to fill up again

Note that this could be a bug in the progressbar itself. A different artifact happens on Wipe device and others: instead of resuming from the exact spot, it resumes _from beginning_. In this case however the response is instanteous. With locking, there is a lag before the progress bar responds.

mmilata commented:
I see, thanks! This should be fixed in e425c15, the progressbar now starts from the beginning as it does in hold-to-confirm dialogs.

Looking into making it resume from the current spot instead of beginning - now that it's on homescreen bored users are gonna play with it and the animation should be as satisfying as possible in this case;)

mmilata commented:
Loader now restarts from the current position and the reverse speed is 3x forward speed.
![Peek 2021-01-27 23-26](https://user-images.githubusercontent.com/85857/106062480-1fb3d580-60f7-11eb-9cb6-5e30f315140c.gif)

matejcik commented on core/src/apps/homescreen/homescreen.py:
```diff
@@ -33,3 +53,39 @@ def on_render(self) -> None:
+    def _loader_start(self) -> None:
+        ui.display.clear()
+        ui.display.text_center(ui.WIDTH // 2, 35, "Hold to lock", ui.BOLD, ui.FG, ui.BG)
+        self.loader.start(_LOADER_DELAY_MS)
```
One more thing, I'm not too convinced about this. It feels jarring that when the loader is shown, it's already a bit up.

mmilata replied:
Removed in 95bbc9e.

tsusanka commented:
I think this is ready? Let's rebase and merge? It looks good to me :+1:

tsusanka commented:
Tested on device and LGTM! Let's merge

mmilata commented:
Added changelog & merged. Thanks for feedbacks & testing!
